### PR TITLE
Feature/deploy-to-ecs

### LINKED ISSUE #675: Migrate Prod API to ECS

### PR DESCRIPTION
**Pull request by dabeeeenster**
todo:

- [x] Update taskRoleArn to include access to AWS secrets
- [x] Update taskRoleArn to include access to dynamodb
- [x]  Update taskRoleArn to include access to ecs:taskRun
- [x] Update taskRoleArn to include access to include events/cloud watch events/event bridge(not sure which specific one I have tested by giving it everything)
- [x] Update container definition to include every environment. 
- [x] Figure out solution to pulumi/pulumi-awsx#856 / https://serverfault.com/questions/1049115/aws-ecs-task-definition-workflow

### CODE REVIEW

File: .github/workflows/api-deploy-production-ecs.yml [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+        steps:
+            - name: Cloning repo
+              uses: actions/checkout@v2
+              with:
+                  fetch-depth: 0
+
+            - name: Configure AWS Credentials
+              uses: aws-actions/configure-aws-credentials@v1
+              with:
+                  aws-access-key-id: ${{ secrets.AWS_ACCESS_KEY_ID }}
```
Comment (matthewelwell) [inline .github/workflows/api-deploy-production-ecs.yml]: Should we move the AWS_ACCESS_KEY_ID to env vars above?

File: .github/workflows/api-deploy-production-ecs.yml [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+              with:
+                  task-definition: infrastructure/aws/production/ecs-task-definition-web.json
+                  container-name: flagsmith-api
+                  image: ${{ steps.build-image.outputs.image }}
+
+            - name: Fill in the new image ID in the Amazon ECS migration task definition
+              id: task-def-api-migration
+              uses: aws-actions/amazon-ecs-render-task-definition@v1
+              with:
+                  task-definition: infrastructure/aws/production/ecs-task-definition-migration.json
```
Comment (matthewelwell) [inline .github/workflows/api-deploy-production-ecs.yml:70]: can we be specific here - is this the db migration task or the identity migration task?

File: .github/workflows/api-deploy-production-ecs.yml [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+                  container-name: flagsmith-api-migration
+                  image: ${{ steps.build-image.outputs.image }}
+
+            - name: Perform migrations
```
Comment (matthewelwell) [inline .github/workflows/api-deploy-production-ecs.yml]: The `name` and `id` are a bit contradictory here, are we registering something or running a task? From the run steps, it looks like we are running a task, can we update the id accordingly?

File: .github/workflows/api-deploy-production-ecs.yml [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+                      "MaximumEventAgeInSeconds": 3600
+                      },
+                      "InputTransformer": {
+                        "InputPathsMap":{"project_id":"$.detail.project_id"},
+                        "InputTemplate": "{ \"containerOverrides\": [ { \"name\": \"flagsmith-api-migration\", \"command\": [\"migrate_identities\", <project_id>], \" environment \":[{\"PROJECT_METADATA_TABLE_NAME_DYNAMO\":\"flagsmith_project_metadata\"}]}]}"
+                        }
+                        }
+                        ]'
+
+#    run-tests:
```
Comment (matthewelwell) [inline .github/workflows/api-deploy-production-ecs.yml:135]: I guess we need to uncomment these before merge?

File: api/scripts/run-docker.sh [hunk 5]
```diff
@@ -6,19 +6,23 @@ function migrate () {
...
+}
+function migrate_identities(){
+    python manage.py migrate_to_edge $1
 }
-
 if [ "$1" == "migrate" ]; then
     migrate
 elif [ "$1" == "serve" ]; then
-    serve
+    serve;
```
Comment (matthewelwell) [inline api/scripts/run-docker.sh]: Can we maintain consistency and either have a trailing ; on all these function calls or none of them?

File: infrastructure/aws/production/ecs-task-definition-migration.json [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+                {
+                    "name": "AWS_REGION",
+                    "value": "eu-west-2"
+                },
+                {
+                    "name": "DJANGO_SETTINGS_MODULE",
+                    "value": "app.settings.production"
+                },
+                {
+                    "name":"IDENTITIES_TABLE_NAME_DYNAMO",
```
Comment (matthewelwell) [inline infrastructure/aws/production/ecs-task-definition-migration.json:22]: Do we need these dynamo tables here? As I understand it, this is the _database_ migration task definition so I don't see why we would need them?

File: infrastructure/aws/production/ecs-task-definition-web.json [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+            "cpu": 0,
+            "portMappings": [
+                {
+                    "containerPort": 8000,
+                    "hostPort": 8000,
+                    "protocol": "tcp"
+                }
+            ],
+            "essential": true,
+            "environment": [
```
Comment (matthewelwell) [inline infrastructure/aws/production/ecs-task-definition-web.json:18]: Surely we need a bunch more env vars here? We have about 30 env vars in beanstalk.

File: .github/workflows/api-deploy-production-ecs.yml (hunk 2 above)
Comment (dabeeeenster) [reply .github/workflows/api-deploy-production-ecs.yml:72]: Renamed

File: .github/workflows/api-deploy-production-ecs.yml (hunk 1 above)
Comment (dabeeeenster) [reply .github/workflows/api-deploy-production-ecs.yml]: Added

File: .github/workflows/api-deploy-production-ecs.yml (hunk 3 above)
Comment (dabeeeenster) [reply .github/workflows/api-deploy-production-ecs.yml]: renamed

File: .github/workflows/api-deploy-production-ecs.yml (hunk 2 above)
Comment (dabeeeenster) [reply .github/workflows/api-deploy-production-ecs.yml:70]: It's both - the image is the same, it's just the CMD parameter sent to the Dockerfile that changes

File: .github/workflows/api-deploy-production-ecs.yml (hunk 4 above)
Comment (dabeeeenster) [reply .github/workflows/api-deploy-production-ecs.yml:135]: Yep will do upon PR Approval

File: infrastructure/aws/production/ecs-task-definition-migration.json (hunk 6 above)
Comment (dabeeeenster) [reply infrastructure/aws/production/ecs-task-definition-migration.json:22]: It's used for both SQL and Identity Migrations

**Review by matthewelwell — APPROVED** [review]

**Pull request merged**